vllm/sequence.py

```diff
@@ -245,21 +246,29 @@ class Sequence:
 
     """
 
-    def __init__(
-            self,
-            seq_id: int,
-            inputs: "LLMInputs",
-            block_size: int,
-            eos_token_id: Optional[int] = None,
-            lora_request: Optional[LoRARequest] = None,
-            prompt_adapter_request: Optional[PromptAdapterRequest] = None
-    ) -> None:
+    def __init__(self,
+                 seq_id: int,
+                 inputs: "LLMInputs",
+                 block_size: int,
+                 eos_token_id: Optional[int] = None,
+                 lora_request: Optional[LoRARequest] = None,
+                 prompt_adapter_request: Optional[PromptAdapterRequest] = None,
+                 from_decoder_prompt: bool = True) -> None:
         self.seq_id = seq_id
         self.inputs = inputs
         self.block_size = block_size
         self.eos_token_id = eos_token_id
         self.lora_request = lora_request
         self.prompt_adapter_request = prompt_adapter_request
+        self.from_decoder_prompt = from_decoder_prompt
+        self._prompt: Optional[str] = None
+        self._prompt_token_ids: Optional[List[int]] = None
+
+        if not (from_decoder_prompt
+                or is_valid_encoder_decoder_llm_inputs(inputs)):
+            raise ValueError("Cannot extract encoder input prompt from "
```

Comment:
This is explained in the new docstring; see below. The True/True scenario you describe is fine, because it is possible to extract a decoder prompt from an LLMInputs instance that has both decoder and encoder prompts (i.e. `is_valid_encoder_decoder_llm_inputs(inputs) == True`.) However the invalid scenario is `from_decoder_prompt == False` and `is_valid_encoder_decoder_llm_inputs(inputs) == False`, i.e. False/False, because you cannot extract an encoder prompt from a decoder-only LLMInputs.

New docstring:

```
"""Stores the data, status, and block information of a sequence.

The sequence is constructed from the LLMInputs instance passed
in through the `inputs` constructor argument.

For encoder/decoder models, LLMInputs encapsulates both a
decoder and encoder prompt, creating an ambiguity about which
prompt to construct the sequence from. The `from_decoder_prompt`
constructor argument signals whether to construct the Sequence
from the LLMInputs decoder prompt, or encoder prompt.

Args:
    seq_id: The ID of the sequence.
    inputs: The inputs of the sequence.
    block_size: The block size of the sequence. Should be the same as the
        block size used by the block manager and cache engine.
    eos_token_id: The end-of-sequence (EOS) token id recognized by this LLM.
    lora_request: LoRA request.
    prompt_adapter_request: Prompt Adapter request.
    from_decoder_prompt: Construct Sequence from LLMInputs decoder prompt
                         (True) or encoder prompt (False.) Must be True
                         for decoder-only model.

"""
```